pymc_experimental/inference/pathfinder/lbfgs.py

```diff
@@ -0,0 +1,126 @@
+import logging
+
+from collections.abc import Callable
+from dataclasses import dataclass, field
+
+import numpy as np
+import pytensor.tensor as pt
+
+from numpy.typing import NDArray
+from pytensor.graph import Apply, Op
+from scipy.optimize import minimize
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(slots=True)
+class LBFGSHistory:
+    x: NDArray[np.float64]
+    g: NDArray[np.float64]
+
+    def __post_init__(self):
+        self.x = np.ascontiguousarray(self.x, dtype=np.float64)
+        self.g = np.ascontiguousarray(self.g, dtype=np.float64)
+
+
+@dataclass(slots=True)
+class LBFGSHistoryManager:
+    fn: Callable[[NDArray[np.float64]], np.float64]
+    grad_fn: Callable[[NDArray[np.float64]], NDArray[np.float64]]
+    x0: NDArray[np.float64]
+    maxiter: int
+    x_history: NDArray[np.float64] = field(init=False)
+    g_history: NDArray[np.float64] = field(init=False)
+    count: int = field(init=False, default=0)
+
+    def __post_init__(self) -> None:
+        self.x_history = np.empty((self.maxiter + 1, self.x0.shape[0]), dtype=np.float64)
+        self.g_history = np.empty((self.maxiter + 1, self.x0.shape[0]), dtype=np.float64)
+
+        value = self.fn(self.x0)
+        grad = self.grad_fn(self.x0)
```

Comment:
You may want to consider a value_and_grad function, as it can avoid many repeated operations?

Reply:
repeated operations means operations that are shared between the logp and dlogp functions. Second approach is fine, except the hardcoded mode, you should allow the user to pass `compile_kwargs` where they can define a custom Mode if they need to